server/src/main/java/org/apache/iotdb/db/qp/logical/crud/QueryOperator.java

```diff
@@ -346,6 +363,21 @@ protected AlignByDevicePlan generateAlignByDevicePlan(PhysicalGenerator generato
     return alignByDevicePlan;
   }
 
+  private void checkWithoutNullColumnValid(AlignByDevicePlan alignByDevicePlan)
+      throws QueryProcessException {
+    if (specialClauseComponent != null) {
+      // meta consistence check
+      for (Expression expression : specialClauseComponent.getWithoutNullColumns()) {
+        if (!alignByDevicePlan.isValidWithoutNullColumn(expression.getExpressionString())) {
+          throw new QueryProcessException(QueryPlan.WITHOUT_NULL_FILTER_ERROR_MESSAGE);
+        }
+      }
+    }
```

Comment:
Honest to say, why don't you combine the check valid method and the calculate index method? They both need to traverse the WithoutNullColumns.

Reply:
fix